tpie/pipelining/node.h

```diff
@@ -573,46 +573,42 @@ class node {
 private:
 	///////////////////////////////////////////////////////////////////////////
 	/// \brief Called by forward_any to add forwarded data.
-	//
-	/// If explicitForward is false, the data will not override data forwarded
-	/// with explicitForward == true.
 	///////////////////////////////////////////////////////////////////////////
-	void add_forwarded_data(std::string key, boost::any value, bool explicitForward);
+	void add_forwarded_data(std::string key, node_token::id_t from_node);
+
+	///////////////////////////////////////////////////////////////////////////
+	/// \brief Called by fetch_any to get data forwarded from this node.
+	///////////////////////////////////////////////////////////////////////////
+	boost::any get_forwarded_data(std::string key, const std::string & type_name = "boost::any");
 
 public:
 	///////////////////////////////////////////////////////////////////////////
 	/// \brief Find out if there is a piece of auxiliary data forwarded with a
 	/// given name.
 	///////////////////////////////////////////////////////////////////////////
 	inline bool can_fetch(std::string key) {
-		return m_values.count(key) != 0;
+		return m_forwardedToHere.count(key) != 0;
 	}
 
 	///////////////////////////////////////////////////////////////////////////
 	/// \brief Fetch piece of auxiliary data as boost::any (the internal
 	/// representation).
 	///////////////////////////////////////////////////////////////////////////
-	boost::any fetch_any(std::string key);
+	boost::any fetch_any(std::string key, const std::string & type_name = "boost::any");
```

Comment:
The typename here is only used for error messages. I do not think it is worth the complications.